 # Random Forest Model

#### eloboration
1. Random forest is an ensemble classifier combining classifier that uses and combining many decision tree
2. ensembling is usually done using the concept of baging with different feature set
3. The reason for using large number of trees in random forest is also to train the trees enough such that contribution from each feature comes in a number of models
4. The result from the ensemble model is usually better than that from the individual decision tree model 

### Strength

1. It runs efficiently on large and expensive data set
2. It has robust method for estimating the missing data and maintain the precision of the data is absent
3. It has powerfull technique for balancing the error in a class and population of unbalanced data sets
4. It gives the estimation about which feature are most important ones on overall classification
5. Lastly random forest algorithm can be used to solve both classification and regression problem

### Weakness of random forest
1. It is not an easy task to understand the random forest as decision tree model
2. It take more computational power than the decision tree

In [1]:
import pandas as pd 
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.ensemble import RandomForestClassifier


In [2]:
# import dataset
data = pd.read_csv('apndcts.csv')

In [3]:
predictors = data.iloc[:,0:7]
predictors

At1    At2    At3    At4    At5    At6    At7
0    0.213  0.554  0.207  0.000  0.000  0.749  0.220
1    0.458  0.714  0.468  0.111  0.102  0.741  0.436
2    0.102  0.518  0.111  0.056  0.022  0.506  0.086
3    0.187  0.196  0.105  0.056  0.029  0.133  0.085
4    0.236  0.804  0.289  0.111  0.066  0.756  0.241
..     ...    ...    ...    ...    ...    ...    ...
101  0.449  0.875  0.523  0.083  0.076  0.920  0.487
102  0.102  0.000  0.022  0.000  0.000  0.000  0.017
103  0.409  0.875  0.482  0.306  0.259  0.914  0.443
104  0.427  0.804  0.474  0.056  0.048  0.836  0.437
105  0.462  0.911  0.551  0.167  0.154  0.931  0.500

[106 rows x 7 columns]

In [4]:
target = data.iloc[:,7]
target

0      1
1      1
2      1
3      1
4      1
      ..
101    0
102    0
103    0
104    0
105    0
Name: class, Length: 106, dtype: int64

In [5]:
predictors_train,predictors_test,target_train,target_test = train_test_split(
    predictors,target,test_size = 0.3,random_state = 123)

In [6]:
rf  = RandomForestClassifier()

In [8]:
# train the model
model = rf.fit(predictors_train,target_train)

In [9]:
# make prediction use the trained model
prediction = rf.predict(predictors_test)

In [10]:
prediction

array([0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0], dtype=int64)

In [11]:
# time to check the prediction accuracy
accuracy_score(target_test,prediction,normalize=True)

0.90625

In [12]:
# This time to check the confusion matrix too
confusion_matrix(target_test,prediction)

array([[25,  0],
       [ 3,  4]], dtype=int64)

Model accuracy = TP + TN / TP+FP+FN+TN
model accracy = 25 + 4/32= 0.90625


sensitivity is model measure the proportion of Positive casse
sensitivity = TP / TP+FN 
sensitivity = 25/28 = 0.8928

###### The high value of sensitivity is more desireable than a high value of accuracy

specify is also another good measure to indicate a good balance of a model being excessvely conservative or excessvely aggresive

specificity = TN/TN+FP 
specificity = 4/4+0 = 1

A higher value of specificity will indicate a better model performance


#### percision
percision gives the proportion of positive predicion which are truly positive
percision = TP/TP+FP
percision = 25/25+0=1=100%

It is quite understand that model higherpercision is precieved to be more reliable

### Recall
Recal the indicate the proportion of correct prediction of positive to the total number of positives

Recall = TP/TP+FN
Recall = 25/28= 0.89

### F-measure
F-measure is another measure of model performance which combine the percision and recall it takes harmonic means of percision and recall

F-measure = 2*percision*recall/percision + recall
F-measure = 2*1*0.89/1+0.89
F-measure = 1.78/1.89
F-measure = 0.9418 = 94%

As a combination of multiple measure into one F-score gives the right  measure using which perfomance of different model can be compared